# 1.What is a parameter?

A parameter is an internal variable that a machine learning model learns from training data.

Example: In a linear regression model y = mx + c, the slope m and intercept c are parameters learned from data.

# 2. What is correlation?

Correlation measures the strength and direction of a linear relationship between two variables.

Value lies between -1 and +1.

+1 → Strong positive relation

-1 → Strong negative relation

0 → No relation

# 3. What does negative correlation mean?

Negative correlation means that when one variable increases, the other decreases.

Example: Number of hours spent watching TV 📺 vs exam score 📊 (more TV, lower score).

# 4. Define Machine Learning. What are the main components in Machine Learning?

Machine Learning (ML) is the science of teaching machines to learn from data and make predictions/decisions without explicit programming.

Main components:

Data – input information

Features – independent variables

Model – algorithm/structure (e.g., regression, decision tree)

Loss function – measures error

Optimizer – reduces loss

Evaluation – checks performance

# 5. How does loss value help in determining whether the model is good or not?

Loss function measures how far predictions are from actual values.

Lower loss → better model.

Example:

MSE (Mean Squared Error) for regression

Cross-Entropy Loss for classification

# 6. What are continuous and categorical variables?

Continuous variable → numeric values with infinite possibilities.
Example: height, weight, temperature.

Categorical variable → represent categories or labels.
Example: gender (male/female), city (Mumbai, Pune).

# 7. How do we handle categorical variables in Machine Learning? What are the common techniques?

Encoding techniques:

Label Encoding – assigns a number to each category (e.g., Male=0, Female=1).

One-Hot Encoding – creates separate binary columns (e.g., Mumbai=[1,0], Pune=[0,1]).

Ordinal Encoding – used when categories have an order (e.g., Small=1, Medium=2, Large=3).

# 8. What do you mean by training and testing a dataset?

Training set → used to teach the model (fit parameters).

Testing set → used to check how well the model generalizes to unseen data.

# 9. What is sklearn.preprocessing?

A module in scikit-learn that provides tools for scaling, normalization, encoding, and transformation of data.

Example: StandardScaler, MinMaxScaler, LabelEncoder, OneHotEncoder.

# 10. What is a Test set?

A portion of data kept aside to evaluate model performance.

Never used during training.

# 11. How do we split data for model fitting (training and testing) in Python?

In [3]:
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
import pandas as pd

# Step 1: Load a sample dataset (Iris dataset)
iris = load_iris()
X = pd.DataFrame(iris.data, columns=iris.feature_names)  # Features
y = pd.Series(iris.target, name='target')               # Target labels

print("X (features):")
display(X.head())
print("\ny (target):")
display(y.head())

# Step 2: Split into Training and Testing sets
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

print("\nTraining set size:", X_train.shape, y_train.shape)
print("Testing set size:", X_test.shape, y_test.shape)


X (features):


sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                5.1               3.5                1.4               0.2
1                4.9               3.0                1.4               0.2
2                4.7               3.2                1.3               0.2
3                4.6               3.1                1.5               0.2
4                5.0               3.6                1.4               0.2


y (target):


0    0
1    0
2    0
3    0
4    0
Name: target, dtype: int64


Training set size: (120, 4) (120,)
Testing set size: (30, 4) (30,)


Explanation:

1. Load dataset

load_iris() loads the famous Iris dataset (150 flower samples).

X = features (sepal & petal measurements).

y = target labels (0 = Setosa, 1 = Versicolor, 2 = Virginica).

2. Convert to Pandas DataFrame/Series

Makes the data easier to view and handle.

X.head() and y.head() show first 5 rows.

3. Split dataset

train_test_split(X, y, test_size=0.2, random_state=42)

80% → Training data (used to train model).

20% → Testing data (used to evaluate model).

random_state=42 ensures reproducibility (same split every time).

4. Check sizes

Training set: 120 samples (80%)

Testing set: 30 samples (20%)

# 12. How do you approach a Machine Learning problem?

Steps:

1. Understand the problem statement

2. Collect & clean data

3. Perform Exploratory Data Analysis (EDA)

4. Feature engineering & encoding

5. Split into train/test

6. Choose ML algorithm

7. Train model

8. Evaluate model (accuracy, precision, recall, etc.)

9. Tune hyperparameters

10. Deploy model

# 13. Why do we have to perform EDA before fitting a model to the data?

EDA helps in:

Understanding data distribution

Detecting outliers & missing values

Identifying correlations

Selecting right features

Without EDA → model may fail or give biased results.

# 16. How can you find correlation between variables in Python?

In [6]:
import pandas as pd
display(X.corr())   # gives correlation matrix

sepal length (cm)  sepal width (cm)  petal length (cm)  \
sepal length (cm)           1.000000         -0.117570           0.871754   
sepal width (cm)           -0.117570          1.000000          -0.428440   
petal length (cm)           0.871754         -0.428440           1.000000   
petal width (cm)            0.817941         -0.366126           0.962865   

                   petal width (cm)  
sepal length (cm)          0.817941  
sepal width (cm)          -0.366126  
petal length (cm)          0.962865  
petal width (cm)           1.000000

# 17. What is causation? Explain difference between correlation and causation with an example.

Causation → one variable directly affects another.

Correlation → two variables move together, but not necessarily cause each other.

Example:

Ice cream sales 🍦 and drowning incidents 🌊 are correlated (both increase in summer).

But ice cream does not cause drowning. → That’s correlation, not causation.

# 18. What is an Optimizer? What are different types of optimizers? Explain each with an example.

Optimizer updates model parameters to minimize loss.

Types:

Gradient Descent – basic form, updates step-by-step.

Stochastic Gradient Descent (SGD) – updates using one sample at a time.

Mini-Batch Gradient Descent – updates with small subsets.

Adam (Adaptive Moment Estimation) – combines momentum + adaptive learning rates.

RMSprop – adjusts learning rate dynamically.

Example in Keras:

In [9]:
from tensorflow.keras.optimizers import Adam
optimizer = Adam(learning_rate=0.001)


# 19. What is sklearn.linear_model ?

A module in scikit-learn with linear models such as:

LinearRegression

LogisticRegression

Ridge, Lasso, ElasticNet

# 20. What does model.fit() do? What arguments must be given?

model.fit(X_train, y_train) → trains the model on training data.

Required arguments:

X_train → input features

y_train → labels/target

#21. What does model.predict() do? What arguments must be given?

model.predict(X_test) → makes predictions on new/unseen data.

Argument: X_test → features (without labels).

# 22. What is feature scaling? How does it help in Machine Learning?

Feature scaling brings variables to the same scale so that no variable dominates others.

Important for algorithms using distance or gradient descent (KNN, SVM, Logistic Regression).

# 23. How do we perform scaling in Python?

In [11]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)


# 26. Explain data encoding?

Data encoding converts categorical data into numerical format for ML algorithms.

Types:

Label Encoding

One-Hot Encoding

Ordinal Encoding

Frequency Encoding